def existing_variant():
    return {
        'Variant_frequency_LOVD': '-',
        'ClinVarAccession_ENIGMA': 'SCV000244373',
        'Variant_in_LOVD': True,
        'Chr': '17',
        'Co_occurrence_LR_exLOVD': '1.23',
        'Reference_Sequence': 'NM_007294.3',
        'Submitter_ClinVar': 'Breast_Cancer_Information_Core_(BIC)_(BRCA1),Ambry_Genetics,GeneDx,Counsyl,Invitae,Evidence-based_Network_for_the_Interpretation_of_Germline_Mutant_Alleles_(ENIGMA),Sharing_Clinical_Reports_Project_(SCRP)',
        'Clinical_classification_BIC': 'Pending',
        'Method_ClinVar': 'curation,clinical_testing,literature_only',
        'Allele_Frequency': '1.647e-05 (ExAC)',
        'Pathogenicity_all': 'Benign(ENIGMA); Benign,Likely_benign,Uncertain_significance (ClinVar); Pending (BIC)',
        'Germline_or_Somatic_BIC': 'G',
        'BIC_Nomenclature': '5074T>C',
        'Clinical_significance_citations_ENIGMA': 'PMID:21990134',
        'Change_Type_id': 1, 'Collection_method_ENIGMA': 'Curation',
        'Sum_family_LR_exLOVD': '0.0182',
        'EAS_Allele_frequency_1000_Genomes': '-',
        'Ethnicity_BIC': 'Western-European,Western European,-,"Western European, Central/Eastern European"',
        'Variant_in_ExAC': True,
        'URL_ENIGMA': 'http://hci-exlovd.hci.utah.edu/variants.php?select_db=BRCA1&action=search_all&search_Variant%2FDNA=c.4955T>C',
        'Allele_Origin_ClinVar': 'germline,unknown',
        'Variant_in_1000_Genomes': False,
        'AFR_Allele_frequency_1000_Genomes': '-',
        'EUR_Allele_frequency_1000_Genomes': '-',
        'Source': 'ENIGMA,ClinVar,LOVD,ExAC,BIC,exLOVD',
        'Condition_ID_value_ENIGMA': 'BREAST-OVARIAN CANCER, FAMILIAL, SUSCEPTIBILITY TO, 1; BROVCA1 (604370)',
        'HGVS_Protein': 'NP_009225.1:p.(Met1652Thr)',
        'Ref': 'A',
        'Sift_Prediction': 'deleterious',
        'Gene_Symbol': 'BRCA1',
        'Comment_on_clinical_significance_ENIGMA': 'IARC class based on posterior probability from multifactorial likelihood analysis, thresholds for class as per Plon et al. 2008 (PMID: 18951446). Class 1 based on posterior probability = 0.00011',
        'Missense_analysis_prior_probability_exLOVD': '0.29',
        'Assertion_method_ENIGMA': 'ENIGMA BRCA1/2 Classification Criteria (2015)',
        'Posterior_probability_exLOVD': '0.00011',
        'Polyphen_Score': '0.546',
        'Hg38_End': 43070959L,
        'HGVS_cDNA': 'NM_007294.3:c.4955T>C',
        'Functional_analysis_technique_LOVD': 'Structure-based prediction (crystallography etc),transcription activation/PS/BA/BS,-,Multifactorial likelihood-ratio model',
        'SAS_Allele_frequency_1000_Genomes': '-',
        'Max_Allele_Frequency': '-',
        'Combined_prior_probablility_exLOVD': '0.29',
        'IARC_class_exLOVD': '1 - Not pathogenic or of no clinical significance',
        'Allele_origin_ENIGMA': 'Germline',
        'Hg36_End': 38476502L,
        'Hg37_End': 41222976L,
        'Date_Last_Updated_ClinVar': '2015-08-10,2014-11-18,2010-06-01,2014-10-01,2002-05-29,2015-04-16,2014-07-23',
        'Hg38_Start': 43070959L,
        'Source_URL': 'http://hci-exlovd.hci.utah.edu/variants.php?select_db=BRCA1&action=search_all&search_Variant%2FDNA=c.4955T>C, http://www.ncbi.nlm.nih.gov/clinvar/?term=SCV000053796, http://www.ncbi.nlm.nih.gov/clinvar/?term=SCV000209981, http://www.ncbi.nlm.nih.gov/clinvar/?term=SCV000220544, http://www.ncbi.nlm.nih.gov/clinvar/?term=SCV000076721, http://www.ncbi.nlm.nih.gov/clinvar/?term=SCV000213478, http://www.ncbi.nlm.nih.gov/clinvar/?term=SCV000145221, http://www.ncbi.nlm.nih.gov/clinvar/?term=SCV000244373',
        'SCV_ClinVar': 'SCV000053796,SCV000209981,SCV000220544,SCV000076721,SCV000213478,SCV000145221,SCV000244373',
        'Pathogenicity_expert': 'Benign / Little Clinical Significance',
        'Allele_frequency_1000_Genomes': '-',
        'Functional_analysis_result_LOVD': 'predicted neutral,predicted deleterious,no functional effect,-',
        'AMR_Allele_frequency_1000_Genomes': '-',
        'Variant_in_ESP': False,
        'Variant_in_BIC': True,
        'Clinical_significance_ENIGMA': 'Benign',
        'Variant_in_ENIGMA': True,
        'Patient_nationality_BIC': 'Not Specified,Western Europe,-',
        'Genomic_Coordinate_hg37': 'chr17:g.41222976:A>G',
        'Genomic_Coordinate_hg36': 'chr17:g.38476502:A>G',
        'Number_of_family_member_carrying_mutation_BIC': '-',
        'Genomic_Coordinate_hg38': 'chr17:g.43070959:A>G',
        'Allele_frequency_ExAC': '1.647e-05',
        'Mutation_type_BIC': 'M',
        'Assertion_method_citation_ENIGMA': 'http://enigmaconsortium.org/documents/ENIGMA_Rules_2015-03-26.pdf',
        'Condition_ID_type_ENIGMA': 'OMIM',
        'Synonyms': '5074T>C,5074T>C,M1652T,NM_007294.2:c.4955T>C,NM_007300.3:c.5018T>C,NM_007299.3:c.1643T>C,NM_007298.3:c.1643T>C,NM_007297.3:c.4814T>C,U14680.1:c.5067T>C',
        'Variant_in_ClinVar': True,
        'Clinical_importance_BIC': 'unknown',
        'Discordant': 'Concordant',
        'Condition_category_ENIGMA': 'Disease',
        'Minor_allele_frequency_percent_ESP': '-',
        'Segregation_LR_exLOVD': '0.012',
        'Variant_in_exLOVD': True,
        'Clinical_Significance_ClinVar': 'Benign,Likely_benign,Uncertain_significance',
        'Protein_Change': 'M1652T',
        'HGVS_RNA': '-',
        'Literature_source_exLOVD': 'Easton DF et al., Am J Hum Genet, 81: 873-883, 2007.',
        'Polyphen_Prediction': 'possibly_damaging',
        'Data_Release_id': 1,
        'Hg37_Start': 41222976L,
        'Hg36_Start': 38476502L,
        'Literature_citation_BIC': '-',
        'Variant_haplotype_LOVD': '-',
        'Pos': '43070959',
        'Date_last_evaluated_ENIGMA': '8/10/15',
        'Sift_Score': '0',
        'Alt': 'G'
    }


def new_variant():
    v = existing_variant()
    v["Genomic_Coordinate_hg38"] = "chr17:g.999999:A>G"
    v['HGVS_cDNA'] = 'NM_007294.3:c.4955T>TC'
    v['BIC_Nomenclature'] = '5074T>TC'
    return v


def new_variant_no_source():
    v = existing_variant()
    v["Genomic_Coordinate_hg38"] = "chr17:g.000000:A>G"
    v['Variant_in_LOVD'] = False
    v['Variant_in_ExAC'] = False
    v['Variant_in_1000_Genomes'] = False
    v['Variant_in_ESP'] = False
    v['Variant_in_BIC'] = False
    v['Variant_in_ENIGMA'] = False
    v['Variant_in_ClinVar'] = False
    v['Variant_in_exLOVD'] = False
    return v


def new_variant_no_enigma_classification():
    v = existing_variant()
    v['Clinical_significance_ENIGMA'] = '-'
    return v
